File: Day 22 : Binary Search Trees/Solution.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node
{
    struct Node* left;
    struct Node* right;
    int data;
}Node;

Node* newNode(int data)
{
    Node* node=(Node*)malloc(sizeof(Node));
    node->left=node->right=NULL;
    node->data=data;
    return node;
}

int max(int x, int y)
{
    if(x > y) return x;
    else return y;
}

int getHeight(Node* root)
{
  if(!root) return -1;
  else if(root == NULL) return 0;
  else
  {
      return (1 + max(getHeight(root -> left), getHeight(root -> right)));    
  }
}

Node* insert(Node* root,int data)
{
    if(root==NULL)
        return newNode(data);
    else
    {
        Node* cur;
        if(data<=root->data)
        {
            cur=insert(root->left,data);
            root->left=cur;                
        }
        else
        {
            cur=insert(root->right,data);
            root->right=cur;
        }
        
    }
    return root;
}

int main()
{
    Node* root=NULL;
    int T,data;
    scanf("%d",&T);
    while(T-->0)
    {
        scanf("%d",&data);
        root=insert(root,data);
    }
    int height=getHeight(root);
    printf("%d",height);
    return 0;   
}

File: Day 22 : Binary Search Trees/Solution.ts
class MyNode {
    data: number
    left: MyNode | null
    right: MyNode | null

    constructor(data: number) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
}

class BinarySearchTree {
    root: MyNode | null = null;

    insert(root: MyNode | null, data: number): MyNode | null {
        if (root === null) {
            const newNode = new MyNode(data);
            if (this.root === null) {
                this.root = newNode;
            }
            return newNode;
        }

        if (data <= root.data) {
            if (root.left) {
                this.insert(root.left, data);
            } else {
                root.left = new MyNode(data);
            }
        } else {
            if (root.right) {
                this.insert(root.right, data);
            } else {
                root.right = new MyNode(data);
            }
        }

        return root;
    }

    getHeight(root: MyNode | null): number {
        if (root === null) {
            return -1;
        }
        return 1 + Math.max(this.getHeight(root.left), this.getHeight(root.right));
    }
}

function main() {
    const tree: BinarySearchTree = new BinarySearchTree();
    let root: MyNode | null = null;

    const n: number = +readLine();
    let valStr: string = "";
    for (let i: number = 0; i < n; i++) {
        valStr += readLine() + " ";
    }
    const values = valStr.split(' ').map(Number);

    for (let i: number = 0; i < values.length; i++) {
        root = tree.insert(root, values[i]);
    }

    console.log(tree.getHeight(root));
}